chainer/functions/math/maximum.py

```diff
@@ -18,24 +19,33 @@ def check_type_forward(self, in_types):
         )
 
     def forward_cpu(self, inputs):
+        self.retain_inputs((0, 1))
         x1, x2 = inputs
         y = numpy.maximum(x1, x2)
         return utils.force_array(y),
 
-    def backward_cpu(self, inputs, grads):
+    def forward_gpu(self, inputs):
+        self.retain_inputs((0, 1))
         x1, x2 = inputs
-        gy, = grads
+        return cuda.cupy.maximum(x1, x2),
+
+    def backward(self, indexes, grad_outputs):
+        x1, x2 = self.get_retained_inputs()
+        return MaximumGrad().apply((x1, x2, grad_outputs[0]))
+
+
+class MaximumGrad(function_node.FunctionNode):
+
+    def forward_cpu(self, inputs):
+        self.retain_inputs((0, 1))
+        x1, x2, gy = inputs
         gx1 = gy * (x1 >= x2)
         gx2 = gy * (x1 < x2)
         return utils.force_array(gx1), utils.force_array(gx2)
 
     def forward_gpu(self, inputs):
-        x1, x2 = inputs
-        return cuda.cupy.maximum(x1, x2),
-
-    def backward_gpu(self, inputs, grads):
-        x1, x2 = inputs
-        gy, = grads
+        self.retain_inputs((0, 1))
+        x1, x2, gy = inputs
         gx1 = cuda.elementwise(
```

Comment:
How about fusing these kernels.

Reply:
Sounds good, but fuse is an experimental feature, so it may raise warning.
Is it acceptable to use fuse in basic features like `max`?